### ISSUE TITLE
add temporaryDirectory() function to Eidos

### ISSUE THREAD

**Issue by bhaller** [MesserLab/SLiM#225]
So, @rdinnager, I'm thinking that I will now add a new function to Eidos, temporaryDirectory(), that returns /tmp/ on non-Windows and whatever it returns on Windows, so that scripts can locate and use the appropriate directory in a cross-platform way.  There are presently five recipes that use /tmp, I think; as part of fixing this issue I'd revise those recipes to use this function.  Does this sound like the right fix to you?

**Comment by bhaller** [comment]
I see it's called tempdir() in R, so I'll use that name.

**Comment by bhaller** [comment]
Along the same lines, @rdinnager, do you think there is any need for an Eidos/SLiM script to be able to tell what platform it is running on?  Should I add an Eidos function of some sort to provide this?  (Looks like R vends this info in various confusing/conflicting ways: https://conjugateprior.org/2015/06/identifying-the-os-from-r/; hard to follow their lead, since Eidos doesn't presently support lists.)

**Comment by rdinnager** [comment]
Sounds good to me. I know all too well how R is pretty bad at this. I used that same function you linked to do this for `slimr` (https://github.com/rdinnager/slimr/blob/83f0cf7077f2703862dfb159a7fea4b51b58d58c/R/utils.R#L24-L50), since I had to detect Windows so that I could use the WSL version of SLiM, which required a different kind of system call to run it. Now that we will have a native Windows build I can take this out of `slimr` now!

**Comment by rdinnager** [comment]
So yes, it might be good to have a function to return the OS. I'm not sure if it will be used that much in SLiM though, but I suppose you never know. If it not too much trouble, might as well have it to fall back on for users if cross-platform issues come up.

**Comment by bhaller** [comment]
OK.  Added `tempdir()` to get the temporary directory in Eidos, and revised recipes that had hard-coded `/tmp` to use `tempdir()` instead.  Also added `sysinfo()` to get various system information; `sysinfo("os")` tells you whether you're on macOS, Windows, or Unix (incl. Linux).  That should fit the bill.  Probably the code in `sysinfo()` will give you a bit of trouble with the port; if you can't find a way to look up equivalent information on Windows, just drop through to the end where it returns `"unknown"`.

BTW, @rdinnager, what do you plan to do about `system()`?  Is there anything to be done, or is it just unsupportable on Windows?  A couple of recipes – 13.5, 14.8 – use it to generate plots from R.  As long as it gives a good error message with `EIDOS_TERMINATE` that's acceptable, although obviously it'd be nice to have a way to make those recipes work in the long run.

**Comment by rdinnager** [comment]
I haven't looked at `system()` yet. I agree it would be good to have it working. In theory it should be possible, I just haven't had a chance to try yet. I will have a look at the new `sysinfo()` function too when I get a chance.

**Comment by rdinnager** [comment]
@bhaller All of `sysinfo()` works on Windows except for the `'login'` and `'user'` options, which I've just `#ifdef`'d out. I can also confirm that `system()` works perfectly on Windows (I only haven't tested the `input` argument, but standard stuff works fine). The Eidos tests are failing because the commands used don't exist in Windows (e.g. `expr`). If you like I can use an `#ifdef` to create Windows versions of those commands in the tests?

**Comment by bhaller** [comment]
> @bhaller All of `sysinfo()` works on Windows except for the `'login'` and `'user'` options, which I've just `#ifdef`'d out. I can also confirm that `system()` works perfectly on Windows (I only haven't tested the `input` argument, but standard stuff works fine). The Eidos tests are failing because the commands used don't exist in Windows (e.g. `expr`). If you like I can use an `#ifdef` to create Windows versions of those commands in the tests?

The `login` option is pretty useless anyway, and `user` was also causing a build failure on Ubuntu 18.04, so I've just removed them both.  As soon as that change passes CI I'll pull it into the Windows branch, so you won't need to `#ifdef` them any more.

Great that `system()` works.  The `expr` usage is just the first thing I thought of to test that `system()` works; of course I have no interest in testing whether `expr` works.  :->  If there's a Windows command accessible through `system()` that also exists on Linux and macOS, that would be ideal.  Something that provides a return value that can be checked for correctness, and something that does not rely on the filesystem or other platform-specific stuff.  Got any ideas for what might work?  If you come up with a good test, post it here and I'll try it on macOS.  :->

**Comment by rdinnager** [comment]
Very few commands on Windows match up with linux commands (I don't know about macOS). Even if the same command works, there is no guarantee they would return the same output. I think `ping localhost` would work on all systems, but the return value would be difficult to check for correctness. I mean `echo` works on Windows, so maybe that is the easiest, just echo a couple of different words and check it returns the word back? I can't think of any other commands that are the same, except `cd`, which obviously is not useful in this case..

**Comment by bhaller** [comment]
@rdinnager the existing tests that use `printf` are working, right?  I think the goal of the `expr` tests is to build towards testing the `stderr` redirection, and I think I can maybe do that with `printf` instead.  What do you get if you execute `printf -xyzzy foo` in your terminal?

**Comment by rdinnager** [comment]
Sorry, no `printf` doesn't work in Windows. As I said, almost no unix commands work in Windows (except if you are using Powershell, which does have more, but `system()` send the CMD.exe, the standard Windows command interpreter). Even Powershell doesn't have `printf` though.

**Comment by bhaller** [comment]
Ah, hmm.  OK.  Sounds like this is indeed a candidate for `#ifdef`, then.  Sounds like you just need to `#ifdef` out all of the `system()` tests except the last one, which uses `echo` and presumably works.  If you can come up with ways to test `system()` more on Windows that would be nice, particularly testing the `args=` and and `input=` and `stderr=` features of it; maybe `echo` can be used for those purposes.  If the tests you come up with happen to work on Un*x systems also, then I'll merge them in, but for now you should leave them `#ifdef`ed for Windows.

I merged in the removal of the `login` and `user` key removal from `sysinfo()` so that's all good.  I think sorting out these failing self-tests is the only thing remaining that would be good to do before I make the first archive on my website for you to try rolling `pacman`, right?  Anything else on the radar on your end?

**Comment by bhaller** [comment]
So, if you start a PR for the self-tests I'll approve that, and then I'll make the archive.  What format should the archive be in, for `pacman`, exactly?

**Comment by rdinnager** [comment]
`tar.gz`
